ModuleTwoFirstTaskDebug/Task1/Product.cs

```diff
@@ -0,0 +1,54 @@
+﻿// <copyright file="Product.cs" company="PlaceholderCompany">
+// Copyright (c) PlaceholderCompany. All rights reserved.
+// </copyright>
+
+namespace ModuleTwoDebugTask
+{
+    /// <summary>
+    /// Some product with it's properties.
+    /// </summary>
+    public class Product
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Product"/> class.
+        /// Sets name and price of the product.
+        /// </summary>
+        /// <param name="name">Name.</param>
+        /// <param name="price">Price.</param>
+        public Product(string name, double price)
+        {
+            this.Name = name;
+            this.Price = price;
+        }
+
+        /// <summary>
+        /// Gets or sets name of product.
+        /// </summary>
+        public string Name { get; set; }
+
+        /// <summary>
+        /// Gets or sets price of product.
+        /// </summary>
+        public double Price { get; set; }
+
+        /// <summary>
+        /// Sets equality rules for the Product type.
+        /// </summary>
+        /// <param name="obj">Product to compare with this instance.</param>
+        /// <returns>Are name and price from input equal to name and price of this instance.</returns>
+        public override bool Equals(object obj)
+        {
+            var product = obj as Product;
+            return product?.Name == this.Name && product.Price == this.Price;
+        }
+
+        /// <summary>
+        /// Overrides standard function to get rid of warning.
+        /// </summary>
+        /// <returns>Hash code.</returns>
+        public override int GetHashCode()
+        {
+            return base.GetHashCode();
```

Comment:
An override of `Equals` implies that you also need to override `GetHashCode` otherwise you may encounter very strange errors in the future.

Please, refactor `GetHashCode` using `System.HashCode` . Also there is an old approach:

```
public override int GetHashCode()
{
    unchecked // Overflow is fine, just wrap
    {
        int hash = 17;
        // Suitable nullity checks etc, of course :)
        hash = hash * 23 + field1.GetHashCode();
        hash = hash * 23 + field2.GetHashCode();
        return hash;
    }
}
```

Reply:
Fixed in a63cd89.